Move important UTxO state transition functions to the top-level and test them

sevanspowell opened the pull request:
This work is intended to allow future modifications to the UTxO state transition function (i.e. collateral) to be tested in isolation, at the unit level.

My changes are primarily to the nested function "applyTx". I have tried to keep my changes minimal:
  - Use more verbose variable names in that function to make the purpose of each variable clear.
  - More clearly delineate between transaction outputs that are unspent and transaction outputs "we know about" but may possibly be spent.
  - Make some effort to separate the ideas of "knowledge of a UTxO" from "ownership of a UTxO" to make testing easier.
  - Extract out the core state transition function from the existing "do lots" "applyTx" function and move it to the top-level.
  - Move other useful functions to the top-level for the same reason.
  - Test important functions in wallet model spec.
  - Re-implement hasKnownInput/hasKnownOutput to make their implementations a little more clear and symmetric.

### Comments

These changes make testing collateral much easier, we can just test the "applyTxToUTxO" function at the unit level, and adjust existing model tests as necessary.

### Issue Number

ADP-1092

sevanspowell committed:
Move important UTxO state transition function to the top-level

- Move the function that modifies the state of a UTxO based on a transaction to
  the top level so that it can be tested.
- Move other useful functions to the top-level for the same reason.
- Test important functions in wallet model.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -320,6 +324,97 @@ totalUTxO
+-- mempty `difference` u = mempty
+-- u `difference` u = mempty
+difference :: UTxO -> UTxO -> UTxO
+difference u1 u2 =
```
**Suggestion**: move this `difference` function to the `UTxO` module.

**Justification**: it's a function that operates purely on `UTxO` values, and could viewed as analogous to these difference functions:

* `TokenBundle.difference`
* `TokenMap.difference`
* `Set.difference`
* `Map.difference`
* ...

sevanspowell replied:
Absolutely :+1:

jonathanknowles replied:
@sevanspowell and I had a discussion about the `difference` function.

We believe it might make more sense if this function were to have the following signature:
```patch
- difference :: UTxO -> UTxO     -> UTxO
+ difference :: UTxO -> Set TxIn -> UTxO
```
Reasons:
- The current implementation relies on finding the difference between two `TxOut` values in the event that a key appears in both UTxO sets.
- But it's not clear what useful answer we can obtain from differencing a pair of `TxOut` values. For example, if the pair of `TxOut` values have different addresses, then which address should we pick, and why?
- In all situations that we currently deal with, there should only be one possible `TxOut` for a given `TxIn`. (We shouldn't ever have a state where there are two possible `TxOut` values for a given `TxIn`.)
- If we assume that there is only one unique mapping for any given `TxIn`, it makes sense to constrain the type so that we're only subtracting a set of `TxIn` values (so the function does not need to deal with the odd case where `TxOut` values differ) from the key set of our `UTxO` parameter.

However, after making this simplification, we can see there is already a pre-existing function with these semantics:
```hs
excluding :: UTxO -> Set TxIn ->  UTxO
excluding (UTxO utxo) =
    UTxO . Map.withoutKeys utxo
```

Hence, in all places that currently use `UTxO.difference`, it might make more sense to call `UTxO.excluding` with just the set of `TxIn` values from the second map.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -320,6 +324,97 @@ totalUTxO
+        filterFunc :: TxIn -> TxOut -> f (Maybe TxOut)
+        filterFunc _txin txout = do
+            ours <- isOursF $ txout ^. #address
+            pure $ if ours then Just txout else Nothing
```
**Suggestion**: Move this function to the `UTxO` module, and perhaps rename it to `filterByAddress` or `filterByAddressM`.

**Justification**: Although this is used in a context where we're filtering for "our" addresses, in principle it could operate with **_any_** monadic indicator function on addresses. Giving it a more general name will match the nice generalization that you've made in the type signature.

jonathanknowles replied:
You could even have a pair of functions:
- `filterByAddressM` (your monadic function above)
- `filterByAddress` (a wrapper around `filterAddressM` that uses the `Identity` monad).

(See here for another example of this pattern: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs#L1847-L1859)

sevanspowell replied:
Yeah, I think that's a good idea. Makes testing easier too, as more often than not the tester will just write the test in terms of some Identity-like monad.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -320,6 +324,97 @@ totalUTxO
+--            ) (outputs tx)
+filterOurUTxOs :: forall f. Monad f => (Address -> f Bool) -> UTxO -> f UTxO
+filterOurUTxOs isOursF (UTxO m) =
+    UTxO <$> Map.traverseMaybeWithKey filterFunc m
```
👍🏻

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -386,16 +481,34 @@ prefilterBlock b u0 = runState $ do
+        -- haven't been validated as "unspent" yet. It is important to not
+        -- remove spent transactions, because if we do, we lose information
+        -- about transactions inputs and outputs we know about.
+        knownTxO <- (prevUTxO <>) <$> filterOurUTxOs isOurs' (utxoFromTx tx)
```
Perhaps a typo?

```suggestion
        knownUTxO <- (prevUTxO <>) <$> filterOurUTxOs isOurs' (utxoFromTx tx)
```

sevanspowell replied:
Not quite. Perhaps you can give me some guidance on this?

My original thought process was that UTxO = **unspent** transaction output, but (something I found quite confusing about this function) is that in this case, `knownTxO` represents transaction outputs we know about and *does not* exclude spent transactions, so it is **not** a UTxO.

Why don't we exclude spent transactions? Because if we do, we lose information about which transactions we "know" about, and so can't calculate "hasKnownInput" and "hasKnownOutput" accurately.

I have created a commit to take a step towards clearing this up.

I have some ideas to clear this up further.

sevanspowell replied:
740c71e

sevanspowell replied:
05d18c6 & 46abf39

sevanspowell committed:
Clear up usage of knownTxO a little

- Remove usage of "knownTxO" when calculating "ourNextUTxO", it's not necessary.
- Move definition of knownTxO closer to where it's actually necessary.

sevanspowell committed:
Clear up language around "knownTxO" a bit more

sevanspowell committed:
Further clear up language around "knownTxO"

sevanspowell committed:
Substitute "difference" with "excluding"

- @jonathanknowles pointed out that the difference function on UTxOs is
  problematic to define over the domain of all UTxOs. Ideally, the mapping
  between TxIn and TxOut should be unique, and so we should never need to worry
  about a case where two different TxOuts are associated with the same TxIn, but
  the existing type can't guarantee this property, and so we need to define the
  behaviour when a TxIn has two different TxOuts. The behaviour is tricky to
  define correctly, so we opt to just use the existing "excluding" function
  instead, which can the same intended function as "difference".

sevanspowell committed:
Move filterOurUTxOs function to UTxO module and rename

- Move the "filterOurUTxOs" function to the UTxO module and rename the function
  to "filterByAddressM".
- Also create a "filterByAddress" function that is the same as
  "filterByAddressM" but runs using the identity functor.
  - Add a test to verify the relationship between the two.

sevanspowell committed:
Fix imports

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs:
```diff
@@ -136,6 +144,29 @@ null (UTxO u) = Map.null u
 size (UTxO u) = Map.size u
 
+-- | Limit a UTxO set to just the UTxOs that are ours, according to some
+-- given function.
```
Suggestion: make the comment more general, to match the generality of the function:

```suggestion
-- | Filter a 'UTxO' set with an indicator function on 'Address' values.
--
-- Returns the subset of UTxO entries that have addresses for which the given
-- indicator function returns 'True'.
```

jonathanknowles replied:
Addressed in https://github.com/input-output-hk/cardano-wallet/pull/2860/commits/7ce039d642e309f206084b012a8b39ee351796db (#2860).

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -386,16 +432,47 @@ prefilterBlock b u0 = runState $ do
+    applyTx (!txs, !prevUTxO) tx = do
+        -- The next UTxO state (apply a state transition) (e.g. remove
+        -- transaction outputs we've spent).
+        ourNextUTxO <- filterByAddressM isOurs' (applyTxToUTxO tx prevUTxO)
```
I suspect this change might alter the performance characteristics of the `applyTx` function.

Before this change:
- We only apply the `isOurs` filter to the outputs of the transaction we're considering.
- We assume that all entries in the current UTxO set already belong to the wallet, so they don't need to be re-filtered.

After this change:
- We first add all transaction outputs to the UTxO set, regardless of whether they belong to us.
- We then have no option but to apply the `isOurs` filter to all entries in the resulting UTxO set, possibly re-filtering entries that have already been filtered before.

I'm wondering, is there a way to rewrite this section so that we don't need to filter the whole UTxO set on each call to `applyTx`?

sevanspowell replied:
Given that `applyTxToUTxO tx u = (u <> utxoFromTx tx) "excluding" inputs tx`, we can work algebraically and re-arrange to: `u "excluding" inputs tx <> utxoFromTx tx`, provided that the law `utxoFromTx tx "excluding" inputs tx === utxoFromTx tx` holds. We definitely want to test it holds, because if it doesn't, it's saying that the UTxO set we produce from the transaction isn't in-fact "unspent" at all!

Re-arranging then helps us to redefine `applyTx` as: `u "excluding" inputs tx <> filterByAdressM isOurs' (utxoFromTx tx)`, which is equivalent to the current implementation, under the assumption that `u` only contains transactions that belong to us.

So if my logic is correct, yes I can re-define it to not filter the whole UTxO set.

sevanspowell replied:
`applyTxToUTxO` is perhaps too course a primitive. I could define a finer primitive `spendTx`, and assert that:
`applyTxToUTxO = spendTx tx u <> utxoFromTx tx = spendTx tx (u <> utxoFromTx tx)`
Then `filterByAddressM isOurs' (applyTxToUTxO tx prevUTxO)`, could be reduced:
```
filterByAddressM isOurs' (applyTxToUTxO tx u)
filterByAddressM isOurs' (spendTx tx u <> utxoFromTx tx) [equivalence mentioned above]
spendTx tx u <> filterByAddressM isOurs' (utxoFromTx tx) [we'd need a new law...]

New law: spendTx tx (filterByAddress f u) = filterByAddress f (spendTx tx u)
```

In my mind, that new law is solid. Spending a tx doesn't change any address properties which could affect a filter. Regardless, the law will constrain the possible implementations.

sevanspowell replied:
This is something Sandy talks about in "Algebra-Driven Design". Defining a program in terms of it's laws and algebraic properties helps us change performance characteristics while maintaining equivalence.

sevanspowell replied:
7717363e7b27c0181e94c183720e0d709b522451

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1431,66 @@ blockchain =
+prop_filterByAddress_allOurs :: Property
+prop_filterByAddress_allOurs =
+    forAllShrink genUTxO shrinkUTxO $ \u ->
+        filterByAddress (const True) u === u
```
👍🏻

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1431,66 @@ blockchain =
+prop_filterByAddress_noneOurs :: Property
+prop_filterByAddress_noneOurs =
+    forAllShrink genUTxO shrinkUTxO $ \u ->
+        filterByAddress (const False) u === mempty
```
👍🏻

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1431,66 @@ blockchain =
+        f = const b
+    in
+        forAllShrink genUTxO shrinkUTxO $ \u ->
+            filterByAddress f u === runIdentity (filterByAddressM (pure . f) u)
```
Is it possible to add a property that uses a real filtering function?

There's already a convenient helper function available we can use here: `addressParity`.

You could choose either of the following:
```hs
(== Even) . addressParity
(==  Odd) . addressParity
```
Either of the above functions will select around half of all addresses.

Then you could verify that all addresses in the resulting UTxO set really do match what you expect.

jonathanknowles replied:
Addressed in https://github.com/input-output-hk/cardano-wallet/pull/2860/commits/d275095bc273a15bf21f117564e8ef80adc8d0bb (#2860).

jonathanknowles committed:
Generalize comments and test names for `filterByAddress`.

The updated comments and test names match the generality of the
function (which can take any indicator function on addresses).

jonathanknowles committed:
Move `UTxO` type specific property tests to `UTxOSpec` module.

jonathanknowles committed:
Move `addressParity` function to `Address.Gen` module.

This makes it reusable from other locations.

We also move `prop_addressParity_coverage` to `AddressSpec`.

jonathanknowles committed:
Add property test `prop_filterByAddress_matchSome`.

sevanspowell committed:
Merge pull request #2860 from input-output-hk/jonathanknowles/adp-1092/apply-tx-testable

Suggestions for PR 2848

sevanspowell committed:
Add spendTx

- Using "applyTxToUTxO" then "filterByAddress" filters the entire UTxO on every
  application of a Tx. By presuming that we've already filtered the previous
  UTxO, then only applying parts of the new Tx that we care about, we can get
  better performance characteristics.
  - Add new "spendTx" function to support this. It does part of what
  "applyTxToUTxO" does.
  - Redefine "applyTxToUTxO" in terms of "spendTx"
  - Replace call to "applyTxToUTxO" with the equivalent, and more efficient:
    "spendTx tx u <> filterByAddress (utxoFromTx tx)"
- Add property tests to prove equivalences.
- Re-jig and clean up existing property tests a bit.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -320,6 +322,53 @@ totalUTxO
+    UTxO $ Map.fromList $ zip (TxIn txId <$> [0..]) outputs
+
+isOurs' :: forall s m. (Monad m, IsOurs s Address) => Address -> StateT s m Bool
+isOurs' addr = do
```
Minor suggestion: we could consider calling this `isOurAddress`, as it's morally equivalent to a specialization of `isOurs` to the `Address` type. (This would make the name a bit more self-documenting, at call sites.)

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -320,6 +322,53 @@ totalUTxO
+    mDerivationIndex <- state $ isOurs addr
+    case mDerivationIndex of
+        Nothing -> pure False
+        Just _  -> pure True
```
```suggestion
isOurs' = fmap isJust . state . isOurs
```

sevanspowell replied:
Done (locally).

sevanspowell committed:
Add commentary and re-arrange property of spendTx

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -386,16 +435,48 @@ prefilterBlock b u0 = runState $ do
+            -- we know about:
+            let known = prevUTxO <> utxoFromTx tx
+            -- But not all those transaction inputs/outputs belong to us:
+            ownedAndKnown <- filterByAddressM isOurs' known
```
Hi Sam. I suspect that evaluating this expression will still re-filter the wallet's whole UTxO, since:
```hs
   filterByAddressM isOurs' known
== filterByAddressM isOurs' (prevUTxO <> utxoFromTx tx)
```
I'm wondering, is there any way we can avoid refiltering parts of the UTxO set that have been filtered before?

Perhaps:
```suggestion
            ownedAndKnown <- (prevUTxO <>)
                <$> filterByAddressM isOurs' (utxoFromTx tx)
```

(Apologies if I've misunderstood something!)

sevanspowell replied:
True! Yep, we can just do what you suggested, I will fix this locally.

sevanspowell committed:
isOurs' -> isOurAddress

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -386,16 +435,48 @@ prefilterBlock b u0 = runState $ do
+        let hasKnownInput =
+                Set.fromList (inputs tx)
+                    `Set.intersection` (Set.fromList ownedAndKnownTxIns)
+                /= mempty
```
Suggestion: use `Set.disjoint`:

```suggestion
        let hasKnownInput = not $ Set.disjoint
                (Set.fromList $ inputs tx)
                (Set.fromList ownedAndKnownTxIns)
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -386,16 +435,48 @@ prefilterBlock b u0 = runState $ do
+        let hasKnownOutput =
+                Set.fromList (outputs tx)
+                    `Set.intersection` (Set.fromList ownedAndKnownTxOuts)
+                /= mempty
```
Suggestion: use `Set.disjoint`:

```suggestion
        let hasKnownOutput = not $ Set.disjoint
                (Set.fromList $ outputs tx)
                (Set.fromList ownedAndKnownTxOuts)
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -451,21 +532,4 @@ changeUTxO
-        state (isOurs $ address out) <&> \case
-            Just{}  -> Just (TxIn (txId tx) ix, out)
-            Nothing -> Nothing
+    mconcat <$> mapM (filterByAddressM isOurs' . utxoFromTx) (Set.toList pending)
```
Nice! 👍🏻

sevanspowell committed:
Make sure not to filter entire UTxO

sevanspowell committed:
Use Set.disjoint

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs:
```diff
@@ -136,6 +144,32 @@ null (UTxO u) = Map.null u
+        filterFunc :: TxIn -> TxOut -> f (Maybe TxOut)
+        filterFunc _txin txout = do
+            ours <- isOursF $ view #address txout
+            pure $ if ours then Just txout else Nothing
```
Formatting:

```suggestion
filterByAddressM isOursF (UTxO m) =
    UTxO <$> Map.traverseMaybeWithKey filterFunc m
  where
    filterFunc :: TxIn -> TxOut -> f (Maybe TxOut)
    filterFunc _txin txout = do
        ours <- isOursF $ view #address txout
        pure $ if ours then Just txout else Nothing
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -182,6 +184,27 @@ spec = do
+
+        describe "utxoFromTx" $ do
+            it "has expected balance" (property prop_utxoFromTx_balance)
+            it "is unspent" (property prop_utxoFromTx_is_unspent)
```
Suggestion: keep to the prevailing style:
```suggestion
        describe "utxoFromTx" $ do
            it "has expected balance"
                (property prop_utxoFromTx_balance)
            it "is unspent"
                (property prop_utxoFromTx_is_unspent)
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -182,6 +184,27 @@ spec = do
+            it "balance is <= balance of UTxO" (property prop_spendTx_balance_inequality)
+            it "has expected balance" (property prop_spendTx_balance)
+            it "definition" (property prop_spendTx)
+            it "commutative with filterByAddress" (property prop_spendTx_filterByAddress)
```
Suggestion: keep to the prevailing style:

```suggestion
        describe "spendTx" $ do
            it "is subset of UTxO"
                (property prop_spendTx_isSubset)
            it "balance is <= balance of UTxO"
                (property prop_spendTx_balance_inequality)
            it "has expected balance"
                (property prop_spendTx_balance)
            it "definition"
                (property prop_spendTx)
            it "commutative with filterByAddress"
                (property prop_spendTx_filterByAddress)
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1372,105 @@ blockchain =
+          `TokenBundle.add`
+              balance (utxoFromTx tx)
+          `TokenBundle.difference`
+              balance (u `UTxO.restrictedBy` Set.fromList (inputs tx))
```
Since we don't modify the generators and shrinkers in any way, perhaps we could define `Arbitrary` instances for `Tx` and `UTxO`, and simplify these properties?

For example:

```suggestion
prop_applyTxToUTxO_balance :: Tx -> UTxO -> Property
prop_applyTxToUTxO_balance tx u =   
    balance (applyTxToUTxO tx u) ===
    balance u
        `TokenBundle.add`
            balance (utxoFromTx tx)
        `TokenBundle.difference`
            balance (u `UTxO.restrictedBy` Set.fromList (inputs tx))
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1372,105 @@ blockchain =
+          `Map.union`
+              unUTxO (utxoFromTx tx)
+          `Map.difference`
+              unUTxO (u `UTxO.restrictedBy` Set.fromList (inputs tx))
```
Suggestion: use `Arbitrary` instances here.

See https://github.com/input-output-hk/cardano-wallet/pull/2848/files#r700738862

sevanspowell committed:
Formatting

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1372,105 @@ blockchain =
+            balance (filterByAddress f (applyTxToUTxO tx mempty))
+            === foldMap (\o -> do
+                    if f (address o) then tokens o else mempty
+                ) (outputs tx)
```
Suggestion: use `Arbitrary` instances here.

See https://github.com/input-output-hk/cardano-wallet/pull/2848/files#r700738862

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1372,105 @@ blockchain =
+prop_utxoFromTx_balance :: Property
+prop_utxoFromTx_balance =
+    forAllShrink genTx shrinkTx $ \tx ->
+        balance (utxoFromTx tx) === foldMap tokens (outputs tx)
```
Suggestion: use `Arbitrary` instances here.

See https://github.com/input-output-hk/cardano-wallet/pull/2848/files#r700738862

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1372,105 @@ blockchain =
+prop_utxoFromTx_is_unspent =
+    forAllShrink genTx shrinkTx $ \tx ->
+        utxoFromTx tx `excluding` Set.fromList (inputs tx)
+        === utxoFromTx tx
```
Suggestion: use `Arbitrary` instances here.

See https://github.com/input-output-hk/cardano-wallet/pull/2848/files#r700738862

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1372,105 @@ blockchain =
+prop_spendTx_isSubset =
+    forAllShrink genTx shrinkTx $ \tx ->
+    forAllShrink genUTxO shrinkUTxO $ \u ->
+        spendTx tx u `UTxO.isSubsetOf` u
```
Suggestion: use `Arbitrary` instances here.

See https://github.com/input-output-hk/cardano-wallet/pull/2848/files#r700738862

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1372,105 @@ blockchain =
+prop_spendTx_balance_inequality :: Property
+prop_spendTx_balance_inequality =
+    forAllShrink genTx shrinkTx $ \tx ->
+    forAllShrink genUTxO shrinkUTxO $ \u ->
```
Suggestion: use `Arbitrary` instances here.

See https://github.com/input-output-hk/cardano-wallet/pull/2848/files#r700738862

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1372,105 @@ blockchain =
+prop_spendTx_balance :: Property
+prop_spendTx_balance =
+    forAllShrink genTx shrinkTx $ \tx ->
+    forAllShrink genUTxO shrinkUTxO $ \u ->
```
Suggestion: use `Arbitrary` instances here.

See https://github.com/input-output-hk/cardano-wallet/pull/2848/files#r700738862

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1372,105 @@ blockchain =
+prop_spendTx :: Property
+prop_spendTx =
+    forAllShrink genTx shrinkTx $ \tx ->
+    forAllShrink genUTxO shrinkUTxO $ \u ->
```
Suggestion: use `Arbitrary` instances here.

See https://github.com/input-output-hk/cardano-wallet/pull/2848/files#r700738862

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1372,105 @@ blockchain =
+prop_applyTxToUTxO_spendTx_utxoFromTx :: Property
+prop_applyTxToUTxO_spendTx_utxoFromTx =
+    forAllShrink genTx shrinkTx $ \tx ->
+    forAllShrink genUTxO shrinkUTxO $ \u ->
```
Suggestion: use `Arbitrary` instances here.

See https://github.com/input-output-hk/cardano-wallet/pull/2848/files#r700738862

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -1407,3 +1372,105 @@ blockchain =
+        f = const b
+    in
+        forAllShrink genTx shrinkTx $ \tx ->
+        forAllShrink genUTxO shrinkUTxO $ \u ->
```
Suggestion: use `Arbitrary` instances here.

See https://github.com/input-output-hk/cardano-wallet/pull/2848/files#r700738862

sevanspowell committed:
Use Arbitrary instances

jonathanknowles reviewed (commented):
Hi Sam

I've already made a few comments. But to add to those:

1. I reckon it would be useful to add some coverage checks to the following properties:
      - [x] `prop_applyTxToUTxO_balance`
      - [x] `prop_applyTxToUTxO_entries`
      - [x] `prop_applyTxToUTxO_spendTx_utxoFromTx`
      - [x] `prop_filterByAddress_balance_applyTxToUTxO`
      - [x] `prop_spendTx_balance_inequality`
      - [x] `prop_spendTx_balance`
      - [x] `prop_spendTx_filterByAddress`
      - [x] `prop_spendTx_isSubset`
      - [x] `prop_spendTx`
      - [x] `prop_utxoFromTx_balance`
      - [x] `prop_utxoFromTx_is_unspent`

1. There are several properties where we use `forAll` with `genX` and `shrinkX` for some type `X`, where we don't modify the generator in any way.

    I think we **_could_** simplify the code a bit by introducing `Arbitrary` instances (within `ModelSpec`) for `UTxO`, `Tx`, and so on. For example: https://github.com/input-output-hk/cardano-wallet/pull/2848/files#r700738862. Though I reckon it's a good idea to do this only after adding coverage checks, in case we need to alter the size parameter for any of the generators.

sevanspowell committed:
Add coverage checks to tests

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -320,6 +322,56 @@ totalUTxO
+    -> UTxO
+applyTxToUTxO tx !u = spendTx tx u <> utxoFromTx tx
+
+-- | Remove unspents that have been consumed by the giventransaction.
```
👾

```suggestion
-- | Remove unspents that have been consumed by the given transaction.
```

sevanspowell replied:
ffedc07

sevanspowell committed:
Fix grammar

jonathanknowles committed:
Use `Arbitrary UTxO` instance to simplify `UTxOSpec`.

jonathanknowles committed:
Add `coarbitraryAddress` function to `Address.Gen`.

This allows the definition of a `CoArbitrary` instance for `Address`,
which makes it possible for QuickCheck to generate arbitrary functions
of type `Address -> t`, provided that type `t` has an `Arbitrary`
instance.

jonathanknowles committed:
Use `CoArbitrary Address` instance to strengthen UTxO property tests.

Instead of using constant functions, we use arbitrary functions of type
`Address -> Bool`.

jonathanknowles committed:
Add coverage checks to `UTxOSpec`.

jonathanknowles committed:
Use `CoArbitrary Address` instance to strengthen `filterByAddress` tests.

Instead of using constant functions, we use arbitrary functions of type
`Address -> Bool`.

jonathanknowles committed:
Make `genTx` sometimes generate transactions with zero outputs.

It's perfectly valid for a transaction to have no outputs if all
incoming value is consumed in some other way (either through the fee, or
by burning tokens).

jonathanknowles committed:
Add coverage checks to individual `ModelSpec` properties.

jonathanknowles committed:
Reduce long execution time of `prop_tx_utxo_coverage`.

By conservatively reducing coverage expectations, this change reduces
the overall run time from around 20 seconds to around 1 second.

jonathanknowles committed:
Strengthen coverage check for `prop_spendTx_balance_inequality`.

What we really want here is to capture the notion of "strictly less
than", but `leq` gives us "less than OR equal". We can't use `<`, as
token maps are only partially ordered. So we have to use a combination
of `leq` and `/=`.

In response to review comment:

https://github.com/input-output-hk/cardano-wallet/pull/2867/files#r701595441

jonathanknowles committed:
Merge pull request #2867 from input-output-hk/jonathanknowles/apply-tx-testable

Strengthen tests for `applyTx` and `filterByAddress`.

jonathanknowles committed:
Merge branch 'master' into sevanspowell/adp-1092/apply-tx-testable

jonathanknowles reviewed (approved):
LGTM!

I've run the test suite with `--qc-max-success 10000`, and all the tests pass.
